### PR TITLE
[DNM] pipeline 2.0: Introducing Data Processing Queue 

### PR DESCRIPTION
**Pull request by marcinszkudlinski**
DP queue is a circular buffer providing safe consumer/producer cached operations cross cores

DP Queue is the only buffering that modules scheduled as DP can use 

Modules can access Dp Queue using sink/source API

Advantages:
 - DP queue does not require any locking - shared part is kept in shared non cached memory (faster than writebacks)
 - DP queue provides cached buffers to modules for faster processing
 - data consumer and producer may safely use it independently at the same time, even from diffrent cores
 - performance: there's only one cache writeback and no additional copying inside

prerequisites:
 1) incoming and outgoing data rate MUST be the same
 2) Both data consumer and data producer declare max chunk sizes
    they want to use (IBS/OBS)
    any requests of bigger data than IBS/OBS will be rejected

 The queue may work in several modes
 1) simple mode
    in case both receiver and sender are located on the
    same core and cache coherency does not matter.
    In this case DP Queue is a simple ring buffer, with IBS and OBS

 2) shared mode
   In this case we need to writeback cache when new
   data arrive and invalidate cache on secondary core.
   That means the whole cacheline must be used exclusively
   by sink or by source, incoming data will be available to use when a
   cacheline is filled completely

TODO - not in this PR - data linearization mode
In this mode the buffers provided to the modules will be linear. 
That will require one extra memcopy operation

### CODE REVIEW

**Comment by cujomalainey** [comment]
@marcinszkudlinski can you leave this as draft until all base PRs have landed?

**Comment by marcinszkudlinski** [comment]
> @marcinszkudlinski can you leave this as draft until all base PRs have landed?

I'd rather mark it as "do not merge" and squash all commit from prev PR to one
Nobody looks at drafts :( and I hope to have some feedback quickly

**Comment by marcinszkudlinski** [comment]
https://github.com/thesofproject/sof/pull/7622 has been merged
rebased on newest main

**Comment by marc-hb** [comment]
> Nobody looks at drafts :( and I hope to have some feedback quickly

I think you're confusing two different things.

The only way to ignore drafts is to explicitly filter them out. Maintainers like @cujomalainey explicitly filter out drafts because they have a LOT of PRs to go through. It's easier for them and their time is rarer and more precious than the time of submitters.

Other people don't filter out drafts. It's true that no one gets notified if you CREATE the PR as a draft the first time (useful feature). BUT if you demote a non-draft to a draft then everyone is _still_ notified. Demoting does NOT unsubscribe.

**Review by kv2019i — COMMENTED** [review]
The need to create queues in adapter in "dp: add dp_queue to module adapter processing " is a bit funky, but the commit describe the intent well and the there is ample inline commentary, so I don't really see blockers with this.

**Comment by lyakh** [comment]
> > @marcinszkudlinski can you leave this as draft until all base PRs have landed?
> 
> I'd rather mark it as "do not merge" and squash all commit from prev PR to one Nobody looks at drafts :( and I hope to have some feedback quickly

@marcinszkudlinski I do look at drafts

**Comment by lyakh** [comment]
@marcinszkudlinski thanks for a nice explanatory text! Could you please explain one more thing: why is our current buffer implementation unsuitable? It's also a ring buffer, so AFAICS it's rather similar to what you're proposing. It *currently* doesn't have all the features that you're describing, but wouldn't it be better to see if those features can be added? Then, also, you're saying that the DP queue can only be used by DP modules. But DP modules are usually linked to "normal" components. What buffers should be used between a DP and a "normal" module? So these buffers are only used if you have at least 2 DP modules linked?

**Comment by kv2019i** [comment]
@lyakh wrote:
> @marcinszkudlinski I do look at drafts. Let's really make it unmergeable somehow - either by making it a draft or I can gladly give it a "-1" just for that ;-) No, not out of any ill intent but we can mark it as -1 if you prefer - just to make accidental merging impossible and let everybody see, that this PR isn't going to be merged tomorrow.

DNM is sufficient to prevent merge. We've used it already to tag PRs that are ready to go review/content wise, but need to wait for some dependency (e.g. Zephyr) to go in. Maybe this warrants a wider discussion (not in this PR), but I do not usually look at drafts and I consider draft-status an indication the PR is not for general consumption. 

This is clearly a PR that is ready for review so I think non-draft is correct status for this.

**Comment by marcinszkudlinski** [comment]
Current implementation may be modified, but I don't see a point to do it - it is already complicated and adding more features may be difficult. The hardest problem I see is buffer aquiring, which is really hard to get rid of because all modules (and other instances) do aquire/release by themselves. 

Sink/src has been introduced to make buffering internals independent from modules, with no additional steps required from the data producers/consumers. That makes possible to create a few specialized buffering implementation instead of a huge unversal one, and some kind of pipeline buffer factory (speaking in C++ language :) ) - modules may provide vector of requirements, factory would chose and configure an implementation that suits both sides - and receive a prepared and ready to use handlers to sink/src. Or similar sollution - to be decided 


As to DP/LL modules:

DP modules must have DP queue on both sides (in/out data), because only DpQueue currently can provide all features required 
for dp processing (cross core, no locking, diffrent IBS/OBS)
 
BUT there are no objections to connect LLs using DpQueue as well. Its just waste of resources as LLs don't need most of the features. In fact a single LL produce data that must be 100% consumed in the same cycle by next LL in pipeline, so shared buffer should be used here (another implementation of sink/source - very very simple one)

The example of pipeline in "2.0 way" may look like

DAI (source) --> **LL1** -->(sink) LL Shared Buffer1(source) -->**LL2** ->(sink)DpQueue(source) --> **DP1** ---> (sink) DpQueue (source)  --> **LL3** -->(sink) DAI 

DpQueue is also a smooth way to connect 2 pipelines running on separate cores

DAI (source) --> **LL1 prim core** -->(sink) LL Shared Buffer1(source) -->**LL2  prim core** ->(sink)DpQueue(source) -->**LL1secCore** ->(sink) LL Shared Buffer2(source) --> **LL2secCore** ->(sink) LL Shared Buffer2(source) --> **LL3secCore** ->(sink) DAI

File: src/audio/dp_queue.c
Comment (lyakh) [inline src/audio/dp_queue.c:124]: isn't free size never greater than `.ibs`? If so, just the second half of the check is enough

File: src/audio/dp_queue.c
Comment (lyakh) [inline src/audio/dp_queue.c:134]: so is the buffer accessed via cached addresses? Then you need `__sparse_cache` for all those addresses. I think both for locking and cache synchronisation the reader will have to do the hard part. But this is the same situation here now as with the current buffer: currently your DP write and read pointers are uncached. So, frequently reading and writing them by both the reader and the writer will be slow. And if you make it cached you'll have to synchronise

File: src/audio/dp_queue.c
Comment (lyakh) [inline src/audio/dp_queue.c:182]: this whole loop should be a single call to `dcache_writeback_invalidate_region()` or 2 of them at most, if wrapping

File: src/audio/dp_queue.c
Comment (lyakh) [inline src/audio/dp_queue.c:172]: so, looks like `.data_buffer` does have to be cached, so a lot of the code has to be changed.

**Review by lyakh — CHANGES_REQUESTED** [review]
I've stopped after the first 2 commits for now, addressing those comments will force changes in the last 2 commits too presumably. If you like you can make a separate PR with the first 2 commits only

File: src/audio/module_adapter/module_adapter.c
Comment (ranj063) [reply src/audio/module_adapter/module_adapter.c:199]: what happens if a modules are disconnected during unbind?

**Review by ranj063 — COMMENTED** [review]
nice PR @marcinszkudlinski. I have the same conern as @lyakh about storing the source/sink acquired pointers and how we deal with module unbind at runtime

File: src/audio/module_adapter/module_adapter.c
Comment (marcinszkudlinski) [reply src/audio/module_adapter/module_adapter.c:199]: in this case I can agree with this.. Will clear all mod->sources[] mod->sinks[] after use. Just using couple of cycles and may sometimes be safer
This is a limitation of audio_stream - as long as it need to be aquired/released. Alternative is to aquire/release each time th halder is used (inside of audio_stream itself). In my opition - too expensive. 

in case of DpQueue the handlers may be kept safely.

File: src/audio/dp_queue.c
Comment (marcinszkudlinski) [reply src/audio/dp_queue.c:18]: I added it to make more clear that this structure contains all dp_q runtime that may safely be cached. Ok, will change

File: src/audio/module_adapter/module_adapter.c
Comment (lyakh) [reply src/audio/module_adapter/module_adapter.c:199]: > This is a limitation of audio_stream - as long as it need to be aquired/released. Alternative is to aquire/release each time th halder is used (inside of audio_stream itself). In my opition - too expensive.

If running on a single core it's cheap - as long as asserts are disabled it's a single pointer conversion when acquiring

> in case of DpQueue the handlers may be kept safely.

Sorry, kept for how long? Not over multiple `.copy()` runs? And in fact DP must be preemptive, so actually they shouldn't be kept for too long anyway

File: src/audio/module_adapter/module_adapter.c
Comment (RanderWang) [inline src/audio/module_adapter/module_adapter.c:475]: "mod->num_of_sources = 0;" ?

**Comment by marcinszkudlinski** [comment]
I'm closing this PR - will split it to smaller parts

**Pull request closed without merging**
